### PR TITLE
Implement training

### PR DESCRIPTION
**Pull request by cschaefer26**
- added trainer and some tensorboard

### CODE REVIEW

File: trainer.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+from dfa.paths import Paths
+from dfa.text import Tokenizer
+from dfa.utils import to_device
+
+
+class Trainer:
+
+    def __init__(self, paths: Paths) -> None:
+        self.paths = paths
+        self.writer = SummaryWriter(log_dir=paths.checkpoint_dir / 'tensorboard', comment='v1')
```
Comment (cfrancesco) [inline trainer.py]: just curious about the comment

File: train.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+from dfa.model import Aligner
+from dfa.paths import Paths
+from dfa.utils import read_config, unpickle_binary
+from trainer import Trainer
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Preprocessing for DeepForcedAligner.')
+    parser.add_argument('--config', '-c', default='config.yaml', help='Points to the config file.')
+    parser.add_argument('--checkpoint', '-cp', default=None, help='Points to the a model file to restore.')
```
Comment (cfrancesco) [inline train.py:13]: How about a flag here instead? whether to train from scratch or load latest from the checkpoint dir if any?

File: trainer.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        symbols = checkpoint['symbols']
+        tokenizer = Tokenizer(symbols)
+        device = torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')
+
+        model = Aligner.from_checkpoint(checkpoint).to(device)
+        optim = Adam(model.parameters())
+        optim.load_state_dict(checkpoint['optim'])
+
+        for g in optim.param_groups:
+            g['lr'] = lr
```
Comment (cfrancesco) [inline trainer.py:45]: why not optim = optim.Adam(model.parameters(), lr=train_params['learning_rate'])?

File: trainer.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+            latest_checkpoint = self.paths.checkpoint_dir / 'latest_model.pt'
+            torch.save({'model': model.state_dict(), 'optim': optim.state_dict(),
+                        'config': config, 'symbols': symbols},
+                       latest_checkpoint)
+
+    def generate_plots(self, model: Aligner, tokenizer: Tokenizer) -> None:
+        model.eval()
+        device = next(model.parameters()).device
+        longest_mel = torch.tensor(self.longest_mel).unsqueeze(0).float().to(device)
+        pred = model(longest_mel)[0].detach().cpu().softmax(dim=-1)
```
Comment (cfrancesco) [inline trainer.py:94]: why using the longest here? for inspecting the worst case?

**Review by cfrancesco — APPROVED** [review]

File: trainer.py (hunk 1 above)
Comment (cschaefer26) [reply trainer.py]: i think its redundant here

File: train.py (hunk 2 above)
Comment (cschaefer26) [reply train.py:13]: Could do, would you think we still need to allow pointing to checkpoints?

File: trainer.py (hunk 3 above)
Comment (cschaefer26) [reply trainer.py:45]: I think it loads the learning rate from the state dict and we have to override it.

File: trainer.py (hunk 4 above)
Comment (cschaefer26) [reply trainer.py:94]: Yeah basically
